File: android/app/src/main/java/com/myshipr/heremap/HereSpeaker.kt
package com.myshipr.heremap

import android.content.Context
import android.media.AudioAttributes
import android.speech.tts.TextToSpeech
import android.util.Log
import java.util.Locale

/**
 * Speaks turn-by-turn guidance.
 *
 * The HERE SDK writes the instruction ("Turn right onto Elm Street") but has no
 * voice of its own — saying it out loud is the app's job. This wraps Android's
 * [TextToSpeech] and is driven straight from the navigator's event listener, so
 * a maneuver is spoken without a round trip through JS.
 *
 * The engine is declared as navigation guidance rather than plain media, which
 * is what makes the system duck music and route the audio to the car over
 * Bluetooth instead of interrupting a call.
 */
internal class HereSpeaker(context: Context) {

    companion object {
        private const val TAG = "HereSpeaker"
        private const val UTTERANCE_ID = "here-guidance"
    }

    private var engine: TextToSpeech? = null
    private var ready = false

    /** Spoken as soon as the engine finishes starting up. */
    private var pending: String? = null

    private var locale: Locale = Locale.US

    /** Muted sessions still receive text; they just do not say it. */
    var enabled: Boolean = true
        set(value) {
            field = value
            if (!value) stop()
        }

    init {
        engine = TextToSpeech(context.applicationContext) { status ->
            if (status != TextToSpeech.SUCCESS) {
                Log.w(TAG, "TextToSpeech unavailable (status $status) — guidance will be silent")
                return@TextToSpeech
            }
            engine?.setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_ASSISTANCE_NAVIGATION_GUIDANCE)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build()
            )
            applyLanguage()
            ready = true
            // The first instruction usually arrives before the engine is up.
            pending?.let { text ->
                pending = null
                speak(text)
            }
        }
    }

    /** @param code a HERE LanguageCode name such as `EN_US`. */
    fun setLanguage(code: String?) {
        locale = parseLocale(code)
        if (ready) applyLanguage()
    }

    /**
     * Says [text], cutting off whatever is still being spoken.
     *
     * Guidance is only useful while it is still true — queueing would leave the
     * driver hearing the previous turn after taking it.
     */
    fun speak(text: String?) {
        if (!enabled || text.isNullOrBlank()) return
        if (!ready) {
            pending = text
            return
        }
        engine?.speak(text, TextToSpeech.QUEUE_FLUSH, null, UTTERANCE_ID)
    }

    fun stop() {
        pending = null
        try {
            engine?.stop()
        } catch (e: Exception) {
            Log.w(TAG, "stop failed: ${e.message}")
        }
    }

    fun shutdown() {
        stop()
        try {
            engine?.shutdown()
        } catch (e: Exception) {
            Log.w(TAG, "shutdown failed: ${e.message}")
        }
        engine = null
        ready = false
    }

    private fun applyLanguage() {
        val result = engine?.setLanguage(locale)
        if (result == TextToSpeech.LANG_MISSING_DATA ||
            result == TextToSpeech.LANG_NOT_SUPPORTED
        ) {
            // The voice data is a per-device download, so this is a normal
            // outcome rather than a fault. English is always present.
            Log.w(TAG, "no voice data for $locale — falling back to ${Locale.US}")
            engine?.setLanguage(Locale.US)
        }
    }

    /** `EN_US` / `en-US` → [Locale]; anything unrecognised falls back to US English. */
    private fun parseLocale(code: String?): Locale {
        if (code.isNullOrBlank()) return Locale.US
        val parts = code.replace('-', '_').split('_')
        return when {
            parts.size >= 2 -> Locale(parts[0].lowercase(), parts[1].uppercase())
            else -> Locale(parts[0].lowercase())
        }
    }
}

File: ios/myshipr/HereSpeaker.swift
import AVFoundation
import Foundation

/// Speaks turn-by-turn guidance.
///
/// The HERE SDK writes the instruction ("Turn right onto Elm Street") but has no
/// voice of its own — saying it out loud is the app's job. This wraps
/// `AVSpeechSynthesizer` and is driven straight from the navigator's event
/// delegate, so a maneuver is spoken without a round trip through JS.
///
/// The audio session is declared as spoken audio that ducks others, which is
/// what turns music down for the instruction instead of stopping it, and keeps
/// guidance audible over a car's Bluetooth connection.
/// `@unchecked Sendable`: every entry point is called from the main thread (the
/// navigation module marshals onto it), and `AVSpeechSynthesizer` is not itself
/// Sendable, so the compiler cannot verify what the call sites guarantee.
final class HereSpeaker: NSObject, @unchecked Sendable {

    private let synthesizer = AVSpeechSynthesizer()
    private var voiceLanguage = "en-US"

    /// Muted sessions still receive text; they just do not say it.
    var enabled: Bool = true {
        didSet { if !enabled { stop() } }
    }

    override init() {
        super.init()
        synthesizer.delegate = self
    }

    /// - Parameter code: a HERE LanguageCode name such as `EN_US`.
    func setLanguage(_ code: String?) {
        voiceLanguage = Self.bcp47(from: code)
    }

    /// Says `text`, cutting off whatever is still being spoken.
    ///
    /// Guidance is only useful while it is still true — queueing would leave the
    /// driver hearing the previous turn after taking it.
    func speak(_ text: String?) {
        guard enabled, let text = text, !text.isEmpty else { return }

        activateSession()
        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }

        let utterance = AVSpeechUtterance(string: text)
        utterance.voice = AVSpeechSynthesisVoice(language: voiceLanguage)
            ?? AVSpeechSynthesisVoice(language: "en-US")
        synthesizer.speak(utterance)
    }

    func stop() {
        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }
        releaseSession()
    }

    // MARK: - Audio session

    private func activateSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            // .duckOthers lowers music for the instruction rather than stopping
            // it; .mixWithOthers keeps this from seizing the session from any
            // other audio the app is running.
            try session.setCategory(
                .playback, mode: .voicePrompt,
                options: [.duckOthers, .mixWithOthers]
            )
            try session.setActive(true)
        } catch {
            NSLog("[HereSpeaker] could not activate the audio session: \(error)")
        }
    }

    /// Handing the session back is what restores music to full volume — without
    /// it everything else stays ducked for the rest of the trip.
    private func releaseSession() {
        do {
            try AVAudioSession.sharedInstance()
                .setActive(false, options: .notifyOthersOnDeactivation)
        } catch {
            // Another part of the app may still hold the session; that is fine.
        }
    }

    /// `EN_US` / `en_US` → `en-US`; anything unrecognised falls back to US English.
    private static func bcp47(from code: String?) -> String {
        guard let code = code, !code.isEmpty else { return "en-US" }
        let parts = code.replacingOccurrences(of: "-", with: "_").split(separator: "_")
        guard let language = parts.first else { return "en-US" }
        if parts.count >= 2 {
            return "\(language.lowercased())-\(parts[1].uppercased())"
        }
        return language.lowercased()
    }
}

extension HereSpeaker: AVSpeechSynthesizerDelegate {
    func speechSynthesizer(
        _ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance
    ) {
        releaseSession()
    }

    func speechSynthesizer(
        _ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance
    ) {
        // Cancels come from speak() replacing an instruction, so the session is
        // about to be used again — leave it active for the incoming utterance.
    }
}
